docarray/utils/apply.py

```diff
@@ -0,0 +1,299 @@
+from contextlib import nullcontext
+from math import ceil
+from multiprocessing.pool import Pool, ThreadPool
+from types import LambdaType
+from typing import Any, Callable, Generator, Optional, TypeVar, Union
+
+from rich.progress import track
+
+from docarray import BaseDocument
+from docarray.array.abstract_array import AnyDocumentArray
+
+T = TypeVar('T', bound=AnyDocumentArray)
+T_doc = TypeVar('T_doc', bound=BaseDocument)
+
+
+def apply(
+    da: T,
+    func: Callable[[T_doc], T_doc],
+    backend: str = 'thread',
+    num_worker: Optional[int] = None,
+    pool: Optional[Union[Pool, ThreadPool]] = None,
+    show_progress: bool = False,
+) -> None:
+    """
+    Apply `func` to every Document of the given DocumentArray in-place while multithreading
+    or multiprocessing.
+
+    EXAMPLE USAGE
+
+    .. code-block:: python
+
+        from docarray import DocumentArray
+        from docarray.documents import Image
+        from docarray.utils.apply import apply
+
+
+        def load_url_to_tensor(img: Image) -> Image:
+            img.tensor = img.url.load()
+            return img
+
+
+        da = DocumentArray[Image]([Image(url='path/to/img.png') for _ in range(100)])
+        apply(
+            da, load_url_to_tensor, backend='thread'
+        )  # threading is usually a good option for IO-bound tasks such as loading an image from url
+
+        for doc in da:
+            assert doc.tensor is not None
+
+    :param da: DocumentArray to apply function to
+    :param func: a function that takes a :class:`BaseDocument` as input and outputs
+        a :class:`BaseDocument`.
+    :param backend: `thread` for multithreading and `process` for multiprocessing.
+        Defaults to `thread`.
+        In general, if `func` is IO-bound then `thread` is a good choice.
+        On the other hand, if `func` is CPU-bound, then you may use `process`.
```

Comment:
I think we should mention that computation that is offloaded to non-python code (e.g. through np/torch/tf) falls under the "IO-bound" category for the purposes of this. Otherwise users would think that they have to use processing when they actually don't